File: src/bin/make_embedding_data.rs
///
/// Copywrite(c) 2025 metanonia
///
/// AI-HUB 건강자료로 부터 encoder를 파인튜팅하기 위한 자료 생성
///

use anyhow::Result;
use glob::glob;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::fs;

#[derive(Debug, Deserialize)]
struct QARecord {
    question: String,
    answer: String,
    q_type: u8,
}

#[derive(Debug, Serialize)]
struct OutputSampleMC {
    question: String,
    positive: String,
    negative: Vec<String>,
    is_negation: bool,
}

#[derive(Debug, Serialize)]
struct OutputSampleNonMC {
    question: String,
    positive: String,
}

fn remove_bom(content: &str) -> &str {
    content.strip_prefix('\u{FEFF}').unwrap_or(content)
}

fn normalize_question(question: &str) -> (String, bool) {
    let negation_patterns = [
        "아닌 것은",
        "아닌것은",
        "아니 것은",
        "틀린 것은",
        "틀린것은",
        "잘못된 것은",
        "잘못된것은",
        "제외한",
        "제외하면",
        "부적절한",
        "옳지 않은",
        "아니다",
        "거리가 먼",
        "관계없는",
    ];

    let is_negative = negation_patterns
        .iter()
        .any(|&pattern| question.contains(pattern));

    if is_negative {
        (format!("[부정질문] {}", question), true)
    } else {
        (question.to_string(), false)
    }
}

fn process_files(
    pattern: &str,
    samples_mc: &mut Vec<OutputSampleMC>,
    samples_nonmc: &mut Vec<OutputSampleNonMC>,
    choice_re: &Regex,
) -> Result<()> {
    for entry in glob(pattern)? {
        if let Ok(path) = entry {
            let content = fs::read_to_string(&path)?;
            let content = remove_bom(&content);
            let qa: QARecord = serde_json::from_str(&content)?;

            if qa.q_type == 1 {
                // 객관식 처리: 선택지 추출
                let q_parts: Vec<&str> = qa.question.splitn(2, '?').collect();
                let stem = q_parts.get(0).unwrap_or(&"").trim().to_string() + "?";
                let rest = q_parts.get(1).unwrap_or(&"");

                // 부정 질문 감지 및 정규화
                let (normalized_stem, is_negation) = normalize_question(&stem);

                // "1) xxx  2) xxx ..." 형태로 추출
                let mut choices = Vec::new();
                for cap in choice_re.split(rest) {
                    let trimmed = cap.trim();
                    if !trimmed.is_empty() {
                        choices.push(trimmed.to_string());
                    }
                }

                // 답 번호와 내용을 분리 (예: "4) 메타콜린..." → "메타콜린...")
                let answer_content = qa.answer.splitn(2, ')').nth(1).unwrap_or("").trim();

                let mut negative = Vec::new();
                let mut positive = String::new();

                if is_negation {
                    // 부정 질문: 정답이 "틀린 것/아닌 것"이므로
                    // 정답은 negative로, 나머지는 positive 후보로
                    for choice in &choices {
                        if choice.contains(answer_content) {
                            // 이것이 "틀린/아닌" 답이므로 실제로는 학습 시 회피해야 할 것
                            negative.push(choice.to_string());
                        } else {
                            // 나머지는 "맞는" 답들
                            if positive.is_empty() {
                                positive = choice.to_string();
                            } else {
                                // 추가 긍정 샘플도 저장 (첫 번째 외 나머지)
                                negative.push(choice.to_string());
                            }
                        }
                    }

                    // 부정 질문의 경우, 실제 오답(정답으로 표시된 것)을 negative의 맨 앞에 배치
                    // 이렇게 하면 학습 시 명시적으로 회피할 수 있음
                    if !negative.is_empty() {
                        negative.rotate_right(1);
                    }
                } else {
                    // 긍정 질문: 기존 로직
                    for choice in &choices {
                        if choice.contains(answer_content) {
                            positive = choice.to_string();
                        } else {
                            negative.push(choice.to_string());
                        }
                    }
                }

                if !positive.is_empty() {
                    samples_mc.push(OutputSampleMC {
                        question: normalized_stem,
                        positive,
                        negative,
                        is_negation,
                    });
                }
            } else {
                // 주관식
                samples_nonmc.push(OutputSampleNonMC {
                    question: qa.question.clone(),
                    positive: qa.answer.clone(),
                });
            }
        }
    }
    Ok(())
}

fn main() -> Result<()> {
    let pattern_train = "data/Training/**/*.json";
    let pattern_val = "data/Validation/**/*.json";

    let mut output_samples_mc_train = Vec::new();
    let mut output_samples_nonmc_train = Vec::new();
    let mut output_samples_mc_val = Vec::new();
    let mut output_samples_nonmc_val = Vec::new();

    let choice_re = Regex::new(r"\d\)\s?")?;

    // Training 데이터 처리
    process_files(
        pattern_train,
        &mut output_samples_mc_train,
        &mut output_samples_nonmc_train,
        &choice_re,
    )?;

    // Validation 데이터 처리
    process_files(
        pattern_val,
        &mut output_samples_mc_val,
        &mut output_samples_nonmc_val,
        &choice_re,
    )?;

    // Training 파일 저장
    let export_mc_train = serde_json::to_string_pretty(&output_samples_mc_train)?;
    fs::write("models/embedding_train_mc.json", export_mc_train)?;

    let export_nonmc_train = serde_json::to_string_pretty(&output_samples_nonmc_train)?;
    fs::write("models/embedding_train_nonmc.json", export_nonmc_train)?;

    // Validation 파일 저장
    let export_mc_val = serde_json::to_string_pretty(&output_samples_mc_val)?;
    fs::write("models/embedding_val_mc.json", export_mc_val)?;

    let export_nonmc_val = serde_json::to_string_pretty(&output_samples_nonmc_val)?;
    fs::write("models/embedding_val_nonmc.json", export_nonmc_val)?;

    // 통계 출력
    let negation_count_train = output_samples_mc_train
        .iter()
        .filter(|s| s.is_negation)
        .count();
    let negation_count_val = output_samples_mc_val
        .iter()
        .filter(|s| s.is_negation)
        .count();

    println!("=== Training 데이터 ===");
    println!(
        "객관식: {}개 (긍정: {}개, 부정: {}개)",
        output_samples_mc_train.len(),
        output_samples_mc_train.len() - negation_count_train,
        negation_count_train
    );
    println!("주관식: {}개", output_samples_nonmc_train.len());

    println!("\n=== Validation 데이터 ===");
    println!(
        "객관식: {}개 (긍정: {}개, 부정: {}개)",
        output_samples_mc_val.len(),
        output_samples_mc_val.len() - negation_count_val,
        negation_count_val
    );
    println!("주관식: {}개", output_samples_nonmc_val.len());

    Ok(())
}

File: make_embedding_data.py
import glob
import json
import re
import os

# 정규 표현식 패턴
choice_re = re.compile(r"\d\)\s?")

# 부정 질문 패턴 목록
negation_patterns = [
    "아닌 것은", "아닌것은", "아니 것은", "틀린 것은", "틀린것은",
    "잘못된 것은", "잘못된것은", "제외한", "제외하면",
    "부적절한", "옳지 않은", "아니다", "거리가 먼", "관계없는"
]

def remove_bom(content):
    """UTF-8 BOM 제거"""
    return content.lstrip('\ufeff')

def normalize_question(question):
    """질문 정규화 및 부정 질문 판단"""
    is_negation = any(pattern in question for pattern in negation_patterns)
    if is_negation:
        return f"[부정질문] {question}", True
    else:
        return question, False

def process_csv_file(path, samples_mc):
    """CSV 파일 처리 (인덱스 기반)"""
    import csv
    with open(path, 'r', encoding='utf-8-sig') as csvfile:
        reader = csv.reader(csvfile)
        next(reader)  # 헤더 건너뛰기

        for row in reader:
            if len(row) < 11:
                continue

            question = row[4].strip()
            a = row[5].strip()
            b = row[6].strip()
            c = row[7].strip()
            d = row[8].strip()
            e = row[9].strip()
            choices = [a, b, c, d, e]
            answer_raw = row[10].strip()

            # answer 변환 (A/B/C/D/E, 1~5 모두 지원)
            answer_upper = answer_raw.upper()
            if answer_upper == "A":
                ans_idx = 0
            elif answer_upper == "B":
                ans_idx = 1
            elif answer_upper == "C":
                ans_idx = 2
            elif answer_upper == "D":
                ans_idx = 3
            elif answer_upper == "E":
                ans_idx = 4
            else:
                try:
                    ans_idx = int(answer_raw) - 1
                except:
                    ans_idx = 0

            if ans_idx < 0 or ans_idx >= len(choices):
                continue

            answer_content = choices[ans_idx]

            normalized_stem, is_negation = normalize_question(question)

            negative = []
            for i, ch in enumerate(choices):
                if i != ans_idx and ch:
                    negative.append(ch)

            positive = answer_content

            # 정답 존재 시만 추가
            if positive:
                samples_mc.append({
                    "question": normalized_stem,
                    "positive": positive,
                    "negative": negative,
                    "is_negation": is_negation
                })

def process_json_files(pattern, samples_mc, samples_nonmc):
    """JSON 파일 처리"""
    for filepath in glob.glob(pattern, recursive=True):
        try:
            with open(filepath, 'r', encoding='utf-8-sig') as f:
                content = f.read()
                content = remove_bom(content)
                qa = json.loads(content)

                question = qa.get('question', '')
                answer = qa.get('answer', '')
                q_type = qa.get('q_type', 0)

                if q_type == 1:
                    # 객관식 처리
                    parts = question.split('?', 1)
                    if len(parts) < 2:
                        continue

                    stem = parts[0].strip() + '?'
                    rest = parts[1]

                    normalized_stem, is_negation = normalize_question(stem)

                    # 선택지 추출
                    choices = []
                    for cap in choice_re.split(rest):
                        trimmed = cap.strip()
                        if trimmed:
                            choices.append(trimmed)

                    # 정답 내용 추출
                    if ')' in answer:
                        answer_content = answer.split(')', 1)[1].strip()
                    else:
                        answer_content = answer.strip()

                    negative = []
                    positive = ""

                    if is_negation:
                        # 부정 질문: 정답이 negative, 나머지 중 하나가 positive
                        for choice in choices:
                            if answer_content in choice:
                                negative.append(choice)
                            else:
                                if not positive:
                                    positive = choice
                                else:
                                    negative.append(choice)

                        # negative 회전 (Rust의 rotate_right(1) 구현)
                        if negative:
                            negative = [negative[-1]] + negative[:-1]
                    else:
                        # 긍정 질문: 정답이 positive, 나머지가 negative
                        for choice in choices:
                            if answer_content in choice:
                                positive = choice
                            else:
                                negative.append(choice)

                    if positive:
                        samples_mc.append({
                            "question": normalized_stem,
                            "positive": positive,
                            "negative": negative,
                            "is_negation": is_negation
                        })
                else:
                    # 주관식
                    samples_nonmc.append({
                        "question": question,
                        "positive": answer
                    })
        except Exception as e:
            print(f"Error processing {filepath}: {e}")
            continue

def main():
    pattern_train_json = "data/Training/**/*.json"
    pattern_val_json = "data/Validation/**/*.json"
    csv_patterns_train = [
        "KorMedMCQA/data/*-train.csv",
        "KorMedMCQA/data/*-dev.csv"
    ]
    csv_patterns_val = ["KorMedMCQA/data/*-test.csv"]

    samples_mc_train = []
    samples_nonmc_train = []
    samples_mc_val = []
    samples_nonmc_val = []

    # Training JSON 처리
    process_json_files(pattern_train_json, samples_mc_train, samples_nonmc_train)
    print(f"step1 {len(samples_mc_train)}")

    # Validation JSON 처리
    process_json_files(pattern_val_json, samples_mc_val, samples_nonmc_val)
    print(f"step2 {len(samples_mc_val)}")

    # Training CSV 처리
    for csv_pattern in csv_patterns_train:
        for filepath in glob.glob(csv_pattern):
            process_csv_file(filepath, samples_mc_train)
    print(f"step3 {len(samples_mc_train)}")

    # Validation CSV 처리
    for csv_pattern in csv_patterns_val:
        for filepath in glob.glob(csv_pattern):
            process_csv_file(filepath, samples_mc_val)
    print(f"step4 {len(samples_mc_val)}")

    # 결과 저장
    os.makedirs("models", exist_ok=True)

    with open("models/embedding_train_mc.json", 'w', encoding='utf-8') as f:
        json.dump(samples_mc_train, f, ensure_ascii=False, indent=2)

    with open("models/embedding_train_nonmc.json", 'w', encoding='utf-8') as f:
        json.dump(samples_nonmc_train, f, ensure_ascii=False, indent=2)

    with open("models/embedding_val_mc.json", 'w', encoding='utf-8') as f:
        json.dump(samples_mc_val, f, ensure_ascii=False, indent=2)

    with open("models/embedding_val_nonmc.json", 'w', encoding='utf-8') as f:
        json.dump(samples_nonmc_val, f, ensure_ascii=False, indent=2)

    # 통계 출력
    negation_train = sum(1 for s in samples_mc_train if s['is_negation'])
    negation_val = sum(1 for s in samples_mc_val if s['is_negation'])

    print("\n=== Training 데이터 ===")
    print(f"객관식: {len(samples_mc_train)}개 (긍정: {len(samples_mc_train)-negation_train}개, 부정: {negation_train}개)")
    print(f"주관식: {len(samples_nonmc_train)}개")

    print("\n=== Validation 데이터 ===")
    print(f"객관식: {len(samples_mc_val)}개 (긍정: {len(samples_mc_val)-negation_val}개, 부정: {negation_val}개)")
    print(f"주관식: {len(samples_nonmc_val)}개")

if __name__ == "__main__":
    main()
